### PR TITLE
Use Buttered Pids

### PR DESCRIPTION
**Pull request by orneryd**
This gives a define for an alternate pid controller specific to Butterflight. It uses the measurement to calculate D instead of the error plus setpoint weight/transition complicated mess that is BeF.

UPDATE:
I made it an CLI/OSD menu option to "Butter your pids" for easier testing in the field.

in the OSD it's 0-1 (OFF/ON)
in the CLI:
```
buttered_pids = ON
Allowed values: OFF, ON
```

targets can opt in to using it by default with:
```
#define USE_BUTTERED_PIDS
```

### CODE REVIEW

File: src/main/flight/pid.c [hunk 1]
```diff
@@ -246,8 +246,10 @@ void pidInitFilters(const pidProfile_t *pidProfile)
 
 static FAST_RAM float Kp[3], Ki[3], Kd[3];
 static FAST_RAM float maxVelocity[3];
+#ifndef BUTTER_PIDS
```
Comment (kidBrazil) [inline src/main/flight/pid.c]: :heart:

**Comment by robert-b** [comment]
you need to be aware that the source of the low d-term values is the quaternion controller.
it needs to much time and is the source for jitter on the time axis.
     dtermSetpointWeight = pidProfile->dtermSetpointWeight / 127.0f;
     relaxFactor = 1.0f / (pidProfile->setpointRelaxRatio / 100.0f);
you added that shit!
i need the FULL d-term!
solved the root cause why this has been added!

**Comment by robert-b** [comment]
ok, your proposal on the the d-term calculation is not what i prefer.

**Comment by robert-b** [comment]
more awful things - from the bf sources:
            // Divide rate change by dT to get differential (ie dr/dt).
            // dT is fixed and calculated from the target PID loop time
            // This is done to avoid DTerm spikes that occur with dynamically
            // calculated deltaT whenever another task causes the PID
            // loop execution to be delayed.

...

**Comment by sUbLiMe-FPV** [comment]
@orneryd 

Wow! This is awesome man! Can't wait to try it out!

I just have one question, I'm not sure I'm reading it right but it looks like you only made the new pid controller work for Helio Spring? What about all the other Butterflight users? Can't we have the goodness too? Pretty please!!

**Comment by orneryd** [comment]
Didn’t mean to close it I meant to comment from my phone and I have fat thumbs. It’s marked as a define for now which means any target can opt-in to the new pid controller.

**Comment by orneryd** [comment]
I'm going to make it a command line option to switch between the old and new pid controllers

**Comment by sUbLiMe-FPV** [comment]
On second thought I changed my mind. This doesn't seem like a good thing. :(

**Comment by orneryd** [comment]
care to elaborate, @sUbLiMe-FPV ?

**Comment by orneryd** [comment]
@sUbLiMe-FPV yes, I have actually. I also put it on a CLI/OSD option so you can choose your pid controller.

**Comment by robert-b** [comment]
mister sublime fpv.
forget what you said.
my onmibus f7 runs at every speed.
NEVER had any issue.

**Comment by kidBrazil** [comment]
Tested this branch extensively today on multiple quads and can confirm - it is buttery

**Comment by orneryd** [comment]
closing this in favor of the odin branch

**Pull request closed without merging**